File: skills/sci-plot/assets/case-packs/rf-0102/triangular_correlation_cluster.py
from pathlib import Path

import matplotlib as mpl
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
from matplotlib.colors import LinearSegmentedColormap
from matplotlib.patches import Patch
from scipy.cluster.hierarchy import dendrogram, leaves_list, linkage
from scipy.spatial.distance import squareform


def resolve_input(script_dir, filename):
    candidates = (script_dir / "data" / filename, script_dir / filename)
    for candidate in candidates:
        if candidate.is_file():
            return candidate
    raise FileNotFoundError(f"Input file not found: {filename}")


def load_inputs(script_dir):
    data_path = resolve_input(script_dir, "data.csv")
    group_path = resolve_input(script_dir, "group.csv")
    data = pd.read_csv(data_path, index_col=0)
    groups = pd.read_csv(group_path, dtype=str, keep_default_na=False)
    if data.shape[0] < 3 or data.shape[1] < 2:
        raise ValueError("data.csv must contain at least three observations and two variables")
    if data.index.has_duplicates or data.columns.has_duplicates:
        raise ValueError("Observation identifiers and variable names must be unique")
    if list(groups.columns) != ["Bile acid", "order"]:
        raise ValueError("group.csv must contain the columns 'Bile acid' and 'order'")
    if groups["Bile acid"].duplicated().any():
        raise ValueError("Every variable must occur exactly once in group.csv")
    if set(groups["Bile acid"]) != set(data.columns):
        missing = sorted(set(data.columns) - set(groups["Bile acid"]))
        extra = sorted(set(groups["Bile acid"]) - set(data.columns))
        raise ValueError(f"Variable/group mismatch; missing={missing}, extra={extra}")
    numeric = data.apply(pd.to_numeric, errors="raise")
    values = numeric.to_numpy(dtype=float)
    if not np.isfinite(values).all():
        raise ValueError("data.csv contains non-finite values")
    if np.any(values < 0):
        raise ValueError("Log transformation requires non-negative measurements")
    return numeric, groups


def calculate_structure(data):
    values = data.to_numpy(dtype=float)
    positives = values[values > 0]
    if positives.size == 0:
        raise ValueError("At least one positive measurement is required")
    minimum_positive = float(positives.min())
    pseudocount = minimum_positive / 2.0
    if not np.isclose(pseudocount, 0.005, rtol=0, atol=1e-12):
        raise ValueError(
            f"Expected a minimum positive value of 0.01, observed {minimum_positive:g}"
        )
    transformed = np.log10(values + pseudocount)
    standard_deviation = transformed.std(axis=0, ddof=1)
    if np.any(standard_deviation == 0):
        constant = data.columns[standard_deviation == 0].tolist()
        raise ValueError(f"Correlation is undefined for constant variables: {constant}")
    correlation = np.corrcoef(transformed, rowvar=False)
    if not np.isfinite(correlation).all():
        raise ValueError("Correlation matrix contains non-finite values")
    if not np.allclose(correlation, correlation.T, atol=1e-12):
        raise ValueError("Correlation matrix is not symmetric")
    distance_component = 1.0 - correlation
    if distance_component.min() < -1e-10 or distance_component.max() > 2.0 + 1e-10:
        raise ValueError("Correlation-derived distance is outside its valid range")
    distance_component = np.maximum(distance_component, 0.0)
    chord_distance = np.sqrt(2.0 * distance_component)
    chord_distance = (chord_distance + chord_distance.T) / 2.0
    np.fill_diagonal(chord_distance, 0.0)
    hierarchy = linkage(
        squareform(chord_distance, checks=True),
        method="average",
        optimal_ordering=False,
    )
    n_variables = correlation.shape[0]
    minimum_leaf = np.empty(2 * n_variables - 1, dtype=int)
    minimum_leaf[:n_variables] = np.arange(n_variables)
    for merge_index in range(n_variables - 1):
        left = int(hierarchy[merge_index, 0])
        right = int(hierarchy[merge_index, 1])
        if minimum_leaf[left] > minimum_leaf[right]:
            hierarchy[merge_index, 0], hierarchy[merge_index, 1] = (
                hierarchy[merge_index, 1],
                hierarchy[merge_index, 0],
            )
            left, right = right, left
        minimum_leaf[n_variables + merge_index] = min(
            minimum_leaf[left], minimum_leaf[right]
        )
    order = leaves_list(hierarchy)
    ordered = correlation[np.ix_(order, order)]
    if not np.allclose(np.diag(ordered), 1.0, atol=1e-10):
        raise ValueError("Correlation diagonal validation failed")
    return pseudocount, correlation, hierarchy, order, ordered


def build_figure(data, groups, pseudocount, hierarchy, order, ordered):
    mpl.rcParams.update(
        {
            "font.family": "DejaVu Sans",
            "font.size": 8,
            "axes.titlesize": 10,
            "axes.labelsize": 8,
            "axes.linewidth": 0.7,
            "xtick.major.width": 0.6,
            "ytick.major.width": 0.6,
            "savefig.facecolor": "white",
        }
    )
    names = data.columns.to_numpy()[order]
    group_lookup = groups.set_index("Bile acid")["order"]
    group_labels = {
        "priConjSum": "Primary conjugated",
        "primarySum": "Primary unconjugated",
        "secConjSum": "Secondary conjugated",
        "secondarySum": "Secondary unconjugated",
        "": "Unassigned",
    }
    group_colors = {
        "Primary conjugated": "#2A9D8F",
        "Primary unconjugated": "#E9C46A",
        "Secondary conjugated": "#457B9D",
        "Secondary unconjugated": "#E76F51",
        "Unassigned": "#B5B7BA",
    }
    unknown_codes = sorted(set(group_lookup) - set(group_labels))
    if unknown_codes:
        raise ValueError(f"Unknown group codes: {unknown_codes}")
    ordered_groups = [group_labels[group_lookup[name]] for name in names]
    group_rgba = np.array(
        [[mpl.colors.to_rgba(group_colors[label]) for label in ordered_groups]]
    )
    n_variables = len(names)
    extent = (0, n_variables * 10, n_variables * 10, 0)
    figure = plt.figure(figsize=(16, 14), constrained_layout=False)
    grid = figure.add_gridspec(
        3,
        2,
        width_ratios=(1, 0.035),
        height_ratios=(2.0, 0.28, 8.5),
        left=0.20,
        right=0.90,
        bottom=0.23,
        top=0.88,
        hspace=0.03,
        wspace=0.05,
    )
    dendrogram_axis = figure.add_subplot(grid[0, 0])
    group_axis = figure.add_subplot(grid[1, 0])
    heat_axis = figure.add_subplot(grid[2, 0])
    color_axis = figure.add_subplot(grid[2, 1])
    dendrogram(
        hierarchy,
        ax=dendrogram_axis,
        no_labels=True,
        color_threshold=0,
        above_threshold_color="#344E41",
        link_color_func=lambda _: "#344E41",
    )
    for collection in dendrogram_axis.collections:
        collection.set_linewidth(0.75)
    dendrogram_axis.set_xlim(0, n_variables * 10)
    dendrogram_axis.set_ylabel("Chord distance", color="#33413B")
    dendrogram_axis.set_xticks([])
    dendrogram_axis.tick_params(axis="y", labelsize=7, colors="#52625B")
    dendrogram_axis.spines[["top", "right", "bottom"]].set_visible(False)
    dendrogram_axis.spines["left"].set_color("#AEB7B2")
    group_axis.imshow(group_rgba, aspect="auto", extent=extent)
    group_axis.set_xlim(0, n_variables * 10)
    group_axis.set_ylim(10, 0)
    group_axis.set_xticks([])
    group_axis.set_yticks([])
    for spine in group_axis.spines.values():
        spine.set_visible(False)
    palette = LinearSegmentedColormap.from_list(
        "correlation",
        ["#254E70", "#79A9C7", "#F7F5F0", "#E49773", "#9D2A32"],
        N=256,
    )
    masked = np.ma.masked_where(
        np.triu(np.ones_like(ordered, dtype=bool), k=1), ordered
    )
    image = heat_axis.imshow(
        masked,
        cmap=palette,
        vmin=-1,
        vmax=1,
        interpolation="nearest",
        aspect="equal",
        extent=extent,
    )
    centers = np.arange(n_variables) * 10 + 5
    heat_axis.set_xticks(centers)
    heat_axis.set_yticks(centers)
    heat_axis.set_xticklabels(
        names,
        rotation=58,
        ha="right",
        va="top",
        rotation_mode="anchor",
        fontsize=5.8,
        color="#26332D",
    )
    heat_axis.set_yticklabels(names, fontsize=5.8, color="#26332D")
    heat_axis.tick_params(axis="both", length=2.5, color="#89958F", pad=2)
    heat_axis.set_xlim(0, n_variables * 10)
    heat_axis.set_ylim(n_variables * 10, 0)
    heat_axis.spines[["top", "right"]].set_visible(False)
    heat_axis.spines[["left", "bottom"]].set_color("#AEB7B2")
    heat_axis.plot(
        [0, n_variables * 10],
        [0, n_variables * 10],
        color="#FFFFFF",
        linewidth=0.5,
        alpha=0.8,
    )
    colorbar = figure.colorbar(image, cax=color_axis, ticks=[-1, -0.5, 0, 0.5, 1])
    colorbar.set_label("Pearson correlation (r)", labelpad=8)
    colorbar.ax.tick_params(labelsize=7, length=2)
    colorbar.outline.set_linewidth(0.6)
    figure.text(
        0.20,
        0.955,
        "Bile acid correlation structure",
        fontsize=22,
        fontweight="bold",
        color="#17251F",
        ha="left",
        va="top",
    )
    figure.text(
        0.20,
        0.922,
        "Lower-triangular correlation matrix with hierarchical clustering",
        fontsize=11,
        color="#52625B",
        ha="left",
        va="top",
    )
    figure.text(
        0.20,
        0.895,
        (
            f"n = {data.shape[0]} observations · Pearson r on log10(x + "
            f"{pseudocount:.3f}) · chord distance √[2(1 − r)] · average linkage"
        ),
        fontsize=8.5,
        color="#68766F",
        ha="left",
        va="top",
    )
    legend_order = [
        "Primary conjugated",
        "Primary unconjugated",
        "Secondary conjugated",
        "Secondary unconjugated",
        "Unassigned",
    ]
    handles = [
        Patch(facecolor=group_colors[label], edgecolor="none", label=label)
        for label in legend_order
    ]
    figure.legend(
        handles=handles,
        loc="upper right",
        bbox_to_anchor=(0.90, 0.925),
        ncol=1,
        frameon=False,
        title="Bile acid class",
        title_fontsize=8,
        fontsize=7.5,
        handlelength=1.4,
        handleheight=0.8,
        labelspacing=0.45,
    )
    figure.text(
        0.20,
        0.06,
        (
            "The 0.005 pseudocount equals half the smallest positive observation (0.01). "
            "The colored strip reports the supplied group annotation. "
            "Variables without a supplied class are shown as Unassigned."
        ),
        fontsize=7.5,
        color="#68766F",
        ha="left",
        va="bottom",
    )
    return figure


def main():
    script_dir = Path(__file__).resolve().parent
    output_path = script_dir / "triangular_correlation_cluster.png"
    data, groups = load_inputs(script_dir)
    pseudocount, correlation, hierarchy, order, ordered = calculate_structure(data)
    figure = build_figure(data, groups, pseudocount, hierarchy, order, ordered)
    figure.savefig(output_path, dpi=300, bbox_inches=None, facecolor="white")
    plt.close(figure)
    if not output_path.is_file() or output_path.stat().st_size <= 0:
        raise OSError(f"Figure was not written: {output_path}")
    unassigned = int((groups["order"] == "").sum())
    print(
        f"Saved: {output_path}\n"
        f"Observations: {data.shape[0]}; variables: {data.shape[1]}; "
        f"unassigned groups: {unassigned}; pseudocount: {pseudocount:.3f}; "
        f"correlation range: [{correlation.min():.3f}, {correlation.max():.3f}]"
    )


if __name__ == "__main__":
    main()

File: skills/sci-plot/assets/case-packs/rf-0102/triangular_correlation_cluster.R
resolve_script_dir <- function() {
  arguments <- commandArgs(trailingOnly = FALSE)
  file_argument <- grep("^--file=", arguments, value = TRUE)
  if (length(file_argument) > 0) {
    return(dirname(normalizePath(sub("^--file=", "", file_argument[[1]]), mustWork = TRUE)))
  }
  frames <- sys.frames()
  if (length(frames) > 0) {
    for (index in rev(seq_along(frames))) {
      source_file <- frames[[index]]$ofile
      if (!is.null(source_file)) {
        return(dirname(normalizePath(source_file, mustWork = TRUE)))
      }
    }
  }
  normalizePath(getwd(), mustWork = TRUE)
}

resolve_input <- function(script_dir, filename) {
  candidates <- c(
    file.path(script_dir, "data", filename),
    file.path(script_dir, filename)
  )
  existing <- candidates[file.exists(candidates)]
  if (length(existing) == 0) {
    stop(sprintf("Input file not found: %s", filename))
  }
  existing[[1]]
}

script_dir <- resolve_script_dir()
output_path <- file.path(script_dir, "triangular_correlation_cluster_r.png")
data_path <- resolve_input(script_dir, "data.csv")
group_path <- resolve_input(script_dir, "group.csv")

if (!requireNamespace("ggplot2", quietly = TRUE)) {
  stop("Package 'ggplot2' is required")
}
if (!requireNamespace("patchwork", quietly = TRUE)) {
  stop("Package 'patchwork' is required")
}

data_frame <- read.csv(
  data_path,
  row.names = 1,
  check.names = FALSE,
  stringsAsFactors = FALSE
)
group_frame <- read.csv(
  group_path,
  check.names = FALSE,
  stringsAsFactors = FALSE,
  na.strings = character()
)

if (nrow(data_frame) < 3 || ncol(data_frame) < 2) {
  stop("data.csv must contain at least three observations and two variables")
}
if (anyDuplicated(rownames(data_frame)) || anyDuplicated(colnames(data_frame))) {
  stop("Observation identifiers and variable names must be unique")
}
if (!identical(names(group_frame), c("Bile acid", "order"))) {
  stop("group.csv must contain the columns 'Bile acid' and 'order'")
}
if (anyDuplicated(group_frame[["Bile acid"]])) {
  stop("Every variable must occur exactly once in group.csv")
}
if (!setequal(group_frame[["Bile acid"]], colnames(data_frame))) {
  missing_names <- setdiff(colnames(data_frame), group_frame[["Bile acid"]])
  extra_names <- setdiff(group_frame[["Bile acid"]], colnames(data_frame))
  stop(sprintf(
    "Variable/group mismatch; missing=%s; extra=%s",
    paste(missing_names, collapse = ", "),
    paste(extra_names, collapse = ", ")
  ))
}

numeric_data <- as.data.frame(
  lapply(data_frame, function(column) {
    converted <- suppressWarnings(as.numeric(column))
    if (any(is.na(converted) & !is.na(column))) {
      stop("data.csv contains non-numeric measurements")
    }
    converted
  }),
  check.names = FALSE
)
rownames(numeric_data) <- rownames(data_frame)
values <- as.matrix(numeric_data)
storage.mode(values) <- "double"

if (any(!is.finite(values))) {
  stop("data.csv contains non-finite values")
}
if (any(values < 0)) {
  stop("Log transformation requires non-negative measurements")
}

positive_values <- values[values > 0]
if (length(positive_values) == 0) {
  stop("At least one positive measurement is required")
}
minimum_positive <- min(positive_values)
pseudocount <- minimum_positive / 2
if (!isTRUE(all.equal(pseudocount, 0.005, tolerance = 1e-12))) {
  stop(sprintf(
    "Expected a minimum positive value of 0.01, observed %g",
    minimum_positive
  ))
}

transformed <- log10(values + pseudocount)
standard_deviation <- apply(transformed, 2, stats::sd)
if (any(standard_deviation == 0)) {
  stop(sprintf(
    "Correlation is undefined for constant variables: %s",
    paste(colnames(values)[standard_deviation == 0], collapse = ", ")
  ))
}

correlation <- stats::cor(transformed, method = "pearson")
if (any(!is.finite(correlation))) {
  stop("Correlation matrix contains non-finite values")
}
if (!isTRUE(all.equal(correlation, t(correlation), tolerance = 1e-12))) {
  stop("Correlation matrix is not symmetric")
}

distance_component <- 1 - correlation
if (min(distance_component) < -1e-10 || max(distance_component) > 2 + 1e-10) {
  stop("Correlation-derived distance is outside its valid range")
}
distance_component[distance_component < 0] <- 0
chord_distance <- sqrt(2 * distance_component)
chord_distance <- (chord_distance + t(chord_distance)) / 2
diag(chord_distance) <- 0
hierarchy <- stats::hclust(stats::as.dist(chord_distance), method = "average")
minimum_leaf <- integer(ncol(values) - 1)
for (merge_index in seq_len(ncol(values) - 1)) {
  children <- hierarchy$merge[merge_index, ]
  child_minimum <- vapply(children, function(child) {
    if (child < 0) {
      return(-child)
    }
    minimum_leaf[[child]]
  }, integer(1))
  if (child_minimum[[1]] > child_minimum[[2]]) {
    hierarchy$merge[merge_index, ] <- rev(children)
    child_minimum <- rev(child_minimum)
  }
  minimum_leaf[[merge_index]] <- min(child_minimum)
}
canonical_order <- function(node) {
  if (node < 0) {
    return(-node)
  }
  c(
    canonical_order(hierarchy$merge[node, 1]),
    canonical_order(hierarchy$merge[node, 2])
  )
}
hierarchy$order <- canonical_order(ncol(values) - 1)
variable_order <- hierarchy$order
ordered_names <- colnames(values)[variable_order]
ordered_correlation <- correlation[variable_order, variable_order, drop = FALSE]

if (!isTRUE(all.equal(
  unname(diag(ordered_correlation)),
  rep(1, ncol(values)),
  tolerance = 1e-10
))) {
  stop("Correlation diagonal validation failed")
}

group_labels <- stats::setNames(
  c(
    "Primary conjugated",
    "Primary unconjugated",
    "Secondary conjugated",
    "Secondary unconjugated",
    "Unassigned"
  ),
  c("priConjSum", "primarySum", "secConjSum", "secondarySum", "")
)
group_colors <- c(
  "Primary conjugated" = "#2A9D8F",
  "Primary unconjugated" = "#E9C46A",
  "Secondary conjugated" = "#457B9D",
  "Secondary unconjugated" = "#E76F51",
  "Unassigned" = "#B5B7BA"
)
unknown_codes <- setdiff(unique(group_frame$order), names(group_labels))
if (length(unknown_codes) > 0) {
  stop(sprintf("Unknown group codes: %s", paste(unknown_codes, collapse = ", ")))
}

group_lookup <- stats::setNames(group_frame$order, group_frame[["Bile acid"]])
ordered_codes <- unname(group_lookup[ordered_names])
ordered_groups <- ifelse(
  ordered_codes == "",
  "Unassigned",
  unname(group_labels[ordered_codes])
)
n_variables <- length(ordered_names)

heat_rows <- which(lower.tri(ordered_correlation, diag = TRUE), arr.ind = TRUE)
heat_data <- data.frame(
  x = heat_rows[, "col"],
  y = heat_rows[, "row"],
  correlation = ordered_correlation[heat_rows]
)
group_data <- data.frame(
  x = seq_len(n_variables),
  y = 1,
  group = factor(ordered_groups, levels = names(group_colors))
)

leaf_positions <- numeric(n_variables)
leaf_positions[hierarchy$order] <- seq_len(n_variables)
node_x <- numeric(n_variables - 1)
segment_rows <- vector("list", (n_variables - 1) * 3)
segment_index <- 1

for (merge_index in seq_len(n_variables - 1)) {
  children <- hierarchy$merge[merge_index, ]
  child_x <- numeric(2)
  child_height <- numeric(2)
  for (side in 1:2) {
    child <- children[[side]]
    if (child < 0) {
      child_x[[side]] <- leaf_positions[[-child]]
      child_height[[side]] <- 0
    } else {
      child_x[[side]] <- node_x[[child]]
      child_height[[side]] <- hierarchy$height[[child]]
    }
  }
  parent_height <- hierarchy$height[[merge_index]]
  node_x[[merge_index]] <- mean(child_x)
  segment_rows[[segment_index]] <- data.frame(
    x = child_x[[1]],
    xend = child_x[[1]],
    y = child_height[[1]],
    yend = parent_height
  )
  segment_rows[[segment_index + 1]] <- data.frame(
    x = child_x[[1]],
    xend = child_x[[2]],
    y = parent_height,
    yend = parent_height
  )
  segment_rows[[segment_index + 2]] <- data.frame(
    x = child_x[[2]],
    xend = child_x[[2]],
    y = child_height[[2]],
    yend = parent_height
  )
  segment_index <- segment_index + 3
}

dendrogram_data <- do.call(rbind, segment_rows)
palette <- c("#254E70", "#79A9C7", "#F7F5F0", "#E49773", "#9D2A32")

dendrogram_plot <- ggplot2::ggplot(dendrogram_data) +
  ggplot2::geom_segment(
    ggplot2::aes(x = x, xend = xend, y = y, yend = yend),
    linewidth = 0.28,
    color = "#344E41",
    lineend = "round"
  ) +
  ggplot2::scale_x_continuous(limits = c(0.5, n_variables + 0.5), expand = c(0, 0)) +
  ggplot2::scale_y_continuous(expand = ggplot2::expansion(mult = c(0, 0.06))) +
  ggplot2::labs(y = "Chord distance") +
  ggplot2::theme_minimal(base_family = "sans", base_size = 8) +
  ggplot2::theme(
    panel.grid = ggplot2::element_blank(),
    axis.title.x = ggplot2::element_blank(),
    axis.text.x = ggplot2::element_blank(),
    axis.ticks.x = ggplot2::element_blank(),
    axis.text.y = ggplot2::element_text(size = 6.5, color = "#52625B"),
    axis.title.y = ggplot2::element_text(size = 7.5, color = "#33413B"),
    plot.margin = ggplot2::margin(2, 76, 0, 0)
  )

group_plot <- ggplot2::ggplot(group_data, ggplot2::aes(x = x, y = y, fill = group)) +
  ggplot2::geom_tile(width = 1, height = 1) +
  ggplot2::scale_fill_manual(values = group_colors, drop = FALSE) +
  ggplot2::scale_x_continuous(limits = c(0.5, n_variables + 0.5), expand = c(0, 0)) +
  ggplot2::scale_y_continuous(limits = c(0.5, 1.5), expand = c(0, 0)) +
  ggplot2::guides(
    fill = ggplot2::guide_legend(
      title = "Bile acid class",
      ncol = 1,
      override.aes = list(width = 0.8, height = 0.8)
    )
  ) +
  ggplot2::theme_void(base_family = "sans") +
  ggplot2::theme(
    legend.position = "right",
    legend.title = ggplot2::element_text(size = 7.5),
    legend.text = ggplot2::element_text(size = 6.8),
    legend.key.height = grid::unit(3.2, "mm"),
    legend.margin = ggplot2::margin(0, 0, 0, 5),
    plot.margin = ggplot2::margin(0, 0, 0, 0)
  )

heatmap_plot <- ggplot2::ggplot(
  heat_data,
  ggplot2::aes(x = x, y = y, fill = correlation)
) +
  ggplot2::geom_tile(width = 1, height = 1) +
  ggplot2::geom_abline(
    intercept = 0,
    slope = 1,
    linewidth = 0.18,
    color = "#FFFFFF",
    alpha = 0.8
  ) +
  ggplot2::scale_fill_gradientn(
    colors = palette,
    values = scales::rescale(c(-1, -0.5, 0, 0.5, 1)),
    limits = c(-1, 1),
    breaks = c(-1, -0.5, 0, 0.5, 1),
    name = "Pearson correlation (r)"
  ) +
  ggplot2::scale_x_continuous(
    limits = c(0.5, n_variables + 0.5),
    breaks = seq_len(n_variables),
    labels = ordered_names,
    expand = c(0, 0)
  ) +
  ggplot2::scale_y_reverse(
    limits = c(n_variables + 0.5, 0.5),
    breaks = seq_len(n_variables),
    labels = ordered_names,
    expand = c(0, 0)
  ) +
  ggplot2::coord_fixed(clip = "off") +
  ggplot2::theme_minimal(base_family = "sans", base_size = 8) +
  ggplot2::theme(
    panel.grid = ggplot2::element_blank(),
    axis.title = ggplot2::element_blank(),
    axis.text.x = ggplot2::element_text(
      angle = 58,
      hjust = 1,
      vjust = 1,
      size = 5.6,
      color = "#26332D"
    ),
    axis.text.y = ggplot2::element_text(size = 5.6, color = "#26332D"),
    axis.ticks = ggplot2::element_line(linewidth = 0.25, color = "#89958F"),
    axis.ticks.length = grid::unit(1.2, "mm"),
    legend.title = ggplot2::element_text(size = 7.5),
    legend.text = ggplot2::element_text(size = 6.8),
    legend.key.height = grid::unit(25, "mm"),
    plot.margin = ggplot2::margin(0, 0, 0, 0)
  )

subtitle <- sprintf(
  "n = %d observations · Pearson r on log10(x + %.3f) · chord distance √[2(1 − r)] · average linkage",
  nrow(values),
  pseudocount
)
caption <- paste(
  "The 0.005 pseudocount equals half the smallest positive observation (0.01).",
  "The colored strip reports the supplied group annotation.",
  "Variables without a supplied class are shown as Unassigned."
)

combined <- (
  dendrogram_plot /
    group_plot /
    heatmap_plot +
    patchwork::plot_layout(heights = c(2.0, 0.30, 8.5), guides = "collect")
) +
  patchwork::plot_annotation(
    title = "Bile acid correlation structure",
    subtitle = paste(
      "Lower-triangular correlation matrix with hierarchical clustering",
      subtitle,
      sep = "\n"
    ),
    caption = caption,
    theme = ggplot2::theme(
      plot.title = ggplot2::element_text(
        family = "sans",
        size = 22,
        face = "bold",
        color = "#17251F",
        hjust = 0
      ),
      plot.subtitle = ggplot2::element_text(
        family = "sans",
        size = 8.5,
        color = "#52625B",
        lineheight = 1.45,
        hjust = 0
      ),
      plot.caption = ggplot2::element_text(
        family = "sans",
        size = 7.5,
        color = "#68766F",
        hjust = 0
      ),
      plot.margin = ggplot2::margin(16, 18, 16, 18)
    )
  ) &
  ggplot2::theme(
    legend.position = "right",
    legend.box = "vertical",
    legend.margin = ggplot2::margin(0, 0, 0, 6)
  )

ggplot2::ggsave(
  filename = output_path,
  plot = combined,
  width = 16,
  height = 14,
  units = "in",
  dpi = 300,
  bg = "white"
)

if (!file.exists(output_path) || file.info(output_path)$size <= 0) {
  stop(sprintf("Figure was not written: %s", output_path))
}

unassigned <- sum(group_frame$order == "")
cat(sprintf(
  paste0(
    "Saved: %s\n",
    "Observations: %d; variables: %d; unassigned groups: %d; ",
    "pseudocount: %.3f; correlation range: [%.3f, %.3f]\n"
  ),
  normalizePath(output_path, mustWork = TRUE),
  nrow(values),
  ncol(values),
  unassigned,
  pseudocount,
  min(correlation),
  max(correlation)
))
